# 2022.01.24 (월)

## Error/Exception Handling



### 디버깅

- branches - 모든 조건이 원하는대로 동작하는지?
- for loops - 반복문에 진입하는지? 원하는 횟수만큼 실행되는지?
- while loops - 종료조건이 제대로 동작하는지? 무한루프에 빠지지 않았는지?
- functions - 함수의 호출이 제대로 되었는지? parameters가 제대로 들어갔는지?




### 에러와 예외

- Syntax Error: 문법 에러, 문제가 발생한 위치를 캐럿기호(^)를 통해 알려줌
  - EOL (End of Line, 괄호 생략 따위), EOF (End of File) 등 도 포함

- Exception: 예외, 실행 중에 감지되는 에러들을 말함
  - ZeroDivision Error: 0으로 나누고자 할 때 발생
  - Name Error: namespace상에 이름이 없는 경우
  - Type Error: 타입이 불일치, argument 누락, argument 갯수 초과 등의 경우
  - Value Error: 타입은 올바르지만 값이 적절하지 않거나 없는 경우 (예: `int('3.5')`)
  - Index Error: 인덱스가 존재하지 않거나 범위를 벗어나는 경우
  - Key Error: 해당 키가 존재하지 않는 경우
  - Module Not Found Error: 존재하지 않는 모듈을 import 하는 경우
  - Import Error: 모듈은 있으나 존재하지 않는 클래스/함수를 가져오는 경우
  - Keyboard Interrupt: 임의로 프로그램을 종료하는 경우
  - Indentation Error: 들여쓰기(indentation)가 적절하지 않는 경우





### 예외 처리

- try문 + except절을 이용하여 예외 처리 가능

  - try문
    - 오류가 발생할 가능성이 있는 코드를 실행
    - 예외가 발생되지 않으면 except 없이 실행 종료
  - except절
    - 예외가 발생하면 except절이 실행
    - 예외 상황을 처리하는 코드를 받아서 적절한 조취를 취함
  - else절
    - try문에서 예외가 발생하지 않으면 실행
  - finally절
    - 예외 발생 여부와 관계없이 항상 실행

  ```python
  # 예시
  try:
      num = input('값을 입력해주세요: ')
      rlt = 100 / int(num)
  except ValueError:
      print('숫자를 입력하세요.')
  except ZeroDivisionError:
      print('0으로 나눌 수 없습니다.')
  else:
      print('계산중입니다...')
      print(rlt)
  finally:
      print('프로그램을 종료합니다.')
  ```



### 예외 발생 시키기

- raise statement: 예외를 강제로 발생
  - `raise <표현식>(<메세지>)`
  - 예시: `raise ValueError('값 에러 발생')`
- assert statement: 예외를 강제로 발생, 일반적으로 상태를 검증하거나 디버깅 용도로 사용
  - `assert <표현식>, <메세지>`
  - 예시: `assert len([1, 2]) == 1, '길이가 1이 아닙니다.'`
